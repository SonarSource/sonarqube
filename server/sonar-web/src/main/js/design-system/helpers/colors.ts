/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { CSSColor } from '../types/theme';

/* eslint-disable no-bitwise, no-mixed-operators */

export function stringToColor(str: string) {
  let hash = 0;

  for (let i = 0; i < str.length; i++) {
    hash = str.charCodeAt(i) + ((hash << 5) - hash);
  }

  let color = '#';

  for (let i = 0; i < 3; i++) {
    const value = (hash >> (i * 8)) & 0xff;
    color += value.toString(16).padStart(2, '0');
  }

  return color;
}

export function isDarkColor(color: string) {
  color = color.substring(1);

  if (color.length === 3) {
    // shortcut notation: #f90
    color = color[0] + color[0] + color[1] + color[1] + color[2] + color[2];
  }

  const rgb = parseInt(color.substring(1), 16);
  const r = (rgb >> 16) & 0xff;
  const g = (rgb >> 8) & 0xff;
  const b = (rgb >> 0) & 0xff;
  const luma = 0.2126 * r + 0.7152 * g + 0.0722 * b;

  return luma < 140;
}

export function getTextColor(background: string, dark = '#222', light = '#fff') {
  return isDarkColor(background) ? light : dark;
}

export function getRGBAString([r, g, b]: Array<number | string>, a?: number | string) {
  return (a !== undefined ? `rgba(${r},${g},${b},${a})` : `rgb(${r},${g},${b})`) as CSSColor;
}
