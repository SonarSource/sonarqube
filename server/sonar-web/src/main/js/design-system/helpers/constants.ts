/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { theme } from 'twin.macro';

export const DEFAULT_LOCALE = 'en';
export const IS_SSR = typeof window === 'undefined';
export const REACT_DOM_CONTAINER = '#content';

export const RULE_STATUSES = ['READY', 'BETA', 'DEPRECATED'];

export const THROTTLE_SCROLL_DELAY = 10;
export const THROTTLE_KEYPRESS_DELAY = 100;

export const DEBOUNCE_DELAY = 250;

export const DEBOUNCE_LONG_DELAY = 1000;

export const DEBOUNCE_SUCCESS_DELAY = 1000;

export const INTERACTIVE_TOOLTIP_DELAY = 0.5;

export const LEAK_PERIOD = 'sonar.leak.period';

export const LEAK_PERIOD_TYPE = 'sonar.leak.period.type';

export const INPUT_SIZES = {
  small: theme('width.input-small'),
  medium: theme('width.input-medium'),
  large: theme('width.input-large'),
  full: theme('width.full'),
  auto: theme('width.auto'),
};

export const LAYOUT_VIEWPORT_MIN_WIDTH = 1280;
export const LAYOUT_VIEWPORT_MAX_WIDTH = 1280;
export const LAYOUT_VIEWPORT_MAX_WIDTH_LARGE = 1680;
export const LAYOUT_MAIN_CONTENT_GUTTER = 60;
export const LAYOUT_SIDEBAR_WIDTH = 240;
export const LAYOUT_SIDEBAR_COLLAPSED_WIDTH = 60;
export const LAYOUT_SIDEBAR_BREAKPOINT = 1320;
export const LAYOUT_BANNER_HEIGHT = 44;
export const LAYOUT_BRANDING_ICON_WIDTH = 198;
export const LAYOUT_FILTERBAR_HEADER = 56;
export const LAYOUT_GLOBAL_NAV_HEIGHT = 52;
export const LAYOUT_PROJECT_NAV_HEIGHT = 108;
export const LAYOUT_LOGO_MARGIN_RIGHT = 45;
export const LAYOUT_LOGO_MAX_HEIGHT = 40;
export const LAYOUT_LOGO_MAX_WIDTH = 150;
export const LAYOUT_FOOTER_HEIGHT = 69;
export const LAYOUT_NOTIFICATIONSBAR_WIDTH = 350;

export const CORE_CONCEPTS_WIDTH = 350;

export const DARK_THEME_ID = 'dark-theme';

export const OPACITY_20_PERCENT = 0.2;

export const OPACITY_75_PERCENT = 0.75;

export const GLOBAL_POPUP_Z_INDEX = 5000;

export const TOAST_AUTOCLOSE_DELAY = 5000;
