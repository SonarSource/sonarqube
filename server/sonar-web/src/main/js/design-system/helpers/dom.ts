/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

/**
 * This function find the point on the target element using the rect coordinat.
 * This point will serve as an anchor for rect to be attached
 *
 * This function assume that rect side will overlap the target facing side. If not
 * the case the point would be outside the target rect. For now we don't need to
 * handle this situation.
 * @param rect
 * @param targetRect
 */
export function findAnchor(rect: DOMRect, targetRect: DOMRect, offset: number) {
  const offestTop = rect.top < targetRect.top ? targetRect.top - rect.top : 0;
  const offestLeft = rect.left < targetRect.left ? targetRect.left - rect.left : 0;

  if (targetRect.right < rect.left) {
    const left = targetRect.right - rect.left - offset / 2;
    const top =
      (Math.min(targetRect.bottom, rect.bottom) - Math.max(targetRect.top, rect.top)) / 2 -
      offset / 2 +
      offestTop;
    const rotate = '0deg';
    const width = -left - offset;

    return { left, top, rotate, width };
  } else if (targetRect.left > rect.right) {
    const left = rect.width + targetRect.left - rect.right + offset / 2;
    const top =
      (Math.min(targetRect.bottom, rect.bottom) - Math.max(targetRect.top, rect.top)) / 2 -
      offset / 2 +
      offestTop;
    const rotate = '180deg';
    const width = left - rect.width;
    return { left, top, rotate, width };
  } else if (targetRect.bottom < rect.top) {
    const left =
      (Math.min(targetRect.right, rect.right) - Math.max(targetRect.left, rect.left)) / 2 -
      offset / 2 +
      offestLeft;
    const top = targetRect.bottom - rect.top - offset / 2;
    const rotate = '90deg';
    const width = -top - offset;
    return { left, top, rotate, width };
  } else if (targetRect.top > rect.bottom) {
    const left =
      (Math.min(targetRect.right, rect.right) - Math.max(targetRect.left, rect.left)) / 2 -
      offset / 2 +
      offestLeft;
    const top = targetRect.top - rect.top - offset / 2;
    const rotate = '-90deg';
    const width = top - rect.height;
    return { left, top, rotate, width };
  }

  // When rectagle overlap
  return { left: 0, top: 0, rotate: '0deg', width: 0 };
}
