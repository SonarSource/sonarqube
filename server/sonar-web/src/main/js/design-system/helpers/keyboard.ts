/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

export enum Key {
  ArrowLeft = 'ArrowLeft',
  ArrowUp = 'ArrowUp',
  ArrowRight = 'ArrowRight',
  ArrowDown = 'ArrowDown',

  Alt = 'Alt',
  Option = 'Option',
  Backspace = 'Backspace',
  CapsLock = 'CapsLock',
  Meta = 'Meta',
  Control = 'Control',
  Command = 'Command',
  Delete = 'Delete',
  End = 'End',
  Enter = 'Enter',
  Escape = 'Escape',
  Home = 'Home',
  PageDown = 'PageDown',
  PageUp = 'PageUp',
  Shift = 'Shift',
  Space = ' ',
  Tab = 'Tab',
  Click = 'Click',
}

export function isShortcut(event: KeyboardEvent): boolean {
  return event.ctrlKey || event.metaKey;
}

const INPUT_TAGS = ['INPUT', 'SELECT', 'TEXTAREA', 'UBCOMMENT'];

export function isInput(event: KeyboardEvent): boolean {
  const { tagName } = event.target as HTMLElement;
  return INPUT_TAGS.includes(tagName);
}

export function isTextarea(
  event: KeyboardEvent,
): event is KeyboardEvent & { target: HTMLTextAreaElement } {
  return event.target instanceof HTMLTextAreaElement;
}
