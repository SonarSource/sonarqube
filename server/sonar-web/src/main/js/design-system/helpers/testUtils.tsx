/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { RenderOptions, RenderResult, render as rtlRender } from '@testing-library/react';
import userEvent, { UserEvent } from '@testing-library/user-event';
import { Options as UserEventsOptions } from '@testing-library/user-event/dist/types/options';
import { InitialEntry } from 'history';
import { identity, kebabCase } from 'lodash';
import React, { PropsWithChildren, ReactNode } from 'react';
import { HelmetProvider } from 'react-helmet-async';
import { IntlProvider, ReactIntlErrorCode } from 'react-intl';
import { MemoryRouter, Route, Routes } from 'react-router-dom';

type RenderResultWithUser = RenderResult & { user: UserEvent };

export function render(
  ui: React.ReactElement,
  options?: RenderOptions,
  userEventOptions?: UserEventsOptions,
): RenderResultWithUser {
  return { ...rtlRender(ui, options), user: userEvent.setup(userEventOptions) };
}

type RenderContextOptions = Omit<RenderOptions, 'wrapper'> & {
  initialEntries?: InitialEntry[];
  userEventOptions?: UserEventsOptions;
};

export function renderWithContext(
  ui: React.ReactElement,
  { userEventOptions, ...options }: RenderContextOptions = {},
) {
  return render(ui, { ...options, wrapper: getContextWrapper() }, userEventOptions);
}

interface RenderRouterOptions {
  additionalRoutes?: ReactNode;
}

export function renderWithRouter(
  ui: React.ReactElement,
  options: RenderContextOptions & RenderRouterOptions = {},
) {
  const { additionalRoutes, userEventOptions, ...renderOptions } = options;

  function RouterWrapper({ children }: React.PropsWithChildren<object>) {
    return (
      <HelmetProvider>
        <IntlWrapper>
          <MemoryRouter>
            <Routes>
              <Route element={children} path="/" />
              {additionalRoutes}
            </Routes>
          </MemoryRouter>
        </IntlWrapper>
      </HelmetProvider>
    );
  }

  return render(ui, { ...renderOptions, wrapper: RouterWrapper }, userEventOptions);
}

function getContextWrapper() {
  return function ContextWrapper({ children }: React.PropsWithChildren<object>) {
    return (
      <HelmetProvider>
        <IntlWrapper>{children}</IntlWrapper>
      </HelmetProvider>
    );
  };
}

export function mockComponent(name: string, transformProps: (props: any) => any = identity) {
  function MockedComponent({ ...props }: PropsWithChildren<any>) {
    return React.createElement('mocked-' + kebabCase(name), transformProps(props));
  }

  MockedComponent.displayName = `mocked(${name})`;
  return MockedComponent;
}

export const debounceTimer = jest
  .fn()
  .mockImplementation((callback: (...args: unknown[]) => void, timeout: number) => {
    let timeoutId: number;

    const debounced = jest.fn((...args: unknown[]) => {
      window.clearTimeout(timeoutId);

      timeoutId = window.setTimeout(() => {
        callback(...args);
      }, timeout);
    });

    (debounced as typeof debounced & { cancel: () => void }).cancel = jest.fn(() => {
      window.clearTimeout(timeoutId);
    });

    return debounced;
  });

export function IntlWrapper({
  children,
  messages = {},
}: {
  children: ReactNode;
  messages?: Record<string, string>;
}) {
  return (
    <IntlProvider
      defaultLocale="en"
      locale="en"
      messages={messages}
      onError={(e) => {
        // ignore missing translations, there are none!
        if (
          e.code !== ReactIntlErrorCode.MISSING_TRANSLATION &&
          e.code !== ReactIntlErrorCode.UNSUPPORTED_FORMATTER
        ) {
          // eslint-disable-next-line no-console
          console.error(e);
        }
      }}
    >
      {children}
    </IntlProvider>
  );
}
