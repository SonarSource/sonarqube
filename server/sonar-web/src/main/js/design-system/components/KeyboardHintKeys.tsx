/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import styled from '@emotion/styled';
import classNames from 'classnames';
import tw from 'twin.macro';
import { themeColor, themeContrast } from '../helpers';
import { Key } from '../helpers/keyboard';
import { TriangleDownIcon, TriangleLeftIcon, TriangleRightIcon, TriangleUpIcon } from './icons';

export const mappedKeys = {
  [Key.Alt]: 'Alt',
  [Key.ArrowDown]: <TriangleDownIcon />,
  [Key.ArrowLeft]: <TriangleLeftIcon />,
  [Key.ArrowRight]: <TriangleRightIcon />,
  [Key.ArrowUp]: <TriangleUpIcon />,
  [Key.Command]: '⌘',
  [Key.Control]: 'Ctrl',
  [Key.Option]: '⌥',
  [Key.Click]: 'click',
};

const NON_KEY_SYMBOLS = ['+', ' '];

export function KeyboardHintKeys({ command }: { command: string }) {
  const keys = command
    .trim()
    .split(' ')
    .map((key, index) => {
      const uniqueKey = `${key}-${index}`;

      if (NON_KEY_SYMBOLS.includes(key)) {
        return <span key={uniqueKey}>{key}</span>;
      }

      const isNonMappedKey = !(
        Object.keys(mappedKeys).includes(key) || Object.values(mappedKeys).includes(key)
      );

      return (
        <KeyBox className={classNames({ 'sw-px-1': isNonMappedKey })} key={uniqueKey}>
          {Object.keys(mappedKeys).includes(key) ? mappedKeys[key as keyof typeof mappedKeys] : key}
        </KeyBox>
      );
    });

  return <div className="sw-flex sw-gap-1">{keys}</div>;
}

export const KeyBox = styled.span`
  ${tw`sw-flex sw-items-center sw-justify-center`}
  ${tw`sw-px-1/2`}
  ${tw`sw-rounded-1/2`}

  background-color: ${themeColor('keyboardHintKey')};
  color: ${themeContrast('keyboardHintKey')};
`;
