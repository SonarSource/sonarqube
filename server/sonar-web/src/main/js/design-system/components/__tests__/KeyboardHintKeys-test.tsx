/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { Key } from '../../helpers/keyboard';
import { render } from '../../helpers/testUtils';
import { FCProps } from '../../types/misc';
import { KeyboardHintKeys, mappedKeys } from '../KeyboardHintKeys';

it.each(Object.keys(mappedKeys))('should render %s', (key) => {
  const { container } = setupWithProps({ command: key });
  expect(container).toMatchSnapshot();
});

it('should render multiple keys', () => {
  const { container } = setupWithProps({ command: `Use Ctrl + ${Key.ArrowUp} ${Key.ArrowDown}` });
  expect(container).toMatchSnapshot();
});

it('should render multiple keys with non-key symbols', () => {
  const { container } = setupWithProps({
    command: `${Key.Control} + ${Key.ArrowDown} ${Key.ArrowUp}`,
  });
  expect(container).toMatchSnapshot();
});

it('should render a default text if no keys match', () => {
  const { container } = setupWithProps({ command: `${Key.Control} + click` });
  expect(container).toMatchSnapshot();
});

function setupWithProps(props: Partial<FCProps<typeof KeyboardHintKeys>> = {}) {
  return render(<KeyboardHintKeys command={`${Key.ArrowUp}`} {...props} />);
}
