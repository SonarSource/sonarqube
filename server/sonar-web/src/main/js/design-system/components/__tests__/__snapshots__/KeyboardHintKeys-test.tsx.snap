// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Alt 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  padding-left: 0.125rem;
  padding-right: 0.125rem;
  border-radius: 0.125rem;
  background-color: rgb(225,230,243);
  color: rgb(62,67,87);
}

<div>
  <div
    class="sw-flex sw-gap-1"
  >
    <span
      class=" emotion-0 emotion-1"
    >
      Alt
    </span>
  </div>
</div>
`;

exports[`should render ArrowDown 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  padding-left: 0.125rem;
  padding-right: 0.125rem;
  border-radius: 0.125rem;
  background-color: rgb(225,230,243);
  color: rgb(62,67,87);
}

<div>
  <div
    class="sw-flex sw-gap-1"
  >
    <span
      class=" emotion-0 emotion-1"
    >
      <svg
        aria-hidden="true"
        class="octicon octicon-triangle-down"
        fill="currentColor"
        focusable="false"
        height="16"
        style="display: inline-block; user-select: none; vertical-align: middle; overflow: visible;"
        viewBox="0 0 16 16"
        width="16"
      >
        <path
          d="m4.427 7.427 3.396 3.396a.25.25 0 0 0 .354 0l3.396-3.396A.25.25 0 0 0 11.396 7H4.604a.25.25 0 0 0-.177.427Z"
        />
      </svg>
    </span>
  </div>
</div>
`;

exports[`should render ArrowLeft 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  padding-left: 0.125rem;
  padding-right: 0.125rem;
  border-radius: 0.125rem;
  background-color: rgb(225,230,243);
  color: rgb(62,67,87);
}

<div>
  <div
    class="sw-flex sw-gap-1"
  >
    <span
      class=" emotion-0 emotion-1"
    >
      <svg
        aria-hidden="true"
        class="octicon octicon-triangle-left"
        fill="currentColor"
        focusable="false"
        height="16"
        style="display: inline-block; user-select: none; vertical-align: middle; overflow: visible;"
        viewBox="0 0 16 16"
        width="16"
      >
        <path
          d="M9.573 4.427 6.177 7.823a.25.25 0 0 0 0 .354l3.396 3.396a.25.25 0 0 0 .427-.177V4.604a.25.25 0 0 0-.427-.177Z"
        />
      </svg>
    </span>
  </div>
</div>
`;

exports[`should render ArrowRight 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  padding-left: 0.125rem;
  padding-right: 0.125rem;
  border-radius: 0.125rem;
  background-color: rgb(225,230,243);
  color: rgb(62,67,87);
}

<div>
  <div
    class="sw-flex sw-gap-1"
  >
    <span
      class=" emotion-0 emotion-1"
    >
      <svg
        aria-hidden="true"
        class="octicon octicon-triangle-right"
        fill="currentColor"
        focusable="false"
        height="16"
        style="display: inline-block; user-select: none; vertical-align: middle; overflow: visible;"
        viewBox="0 0 16 16"
        width="16"
      >
        <path
          d="m6.427 4.427 3.396 3.396a.25.25 0 0 1 0 .354l-3.396 3.396A.25.25 0 0 1 6 11.396V4.604a.25.25 0 0 1 .427-.177Z"
        />
      </svg>
    </span>
  </div>
</div>
`;

exports[`should render ArrowUp 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  padding-left: 0.125rem;
  padding-right: 0.125rem;
  border-radius: 0.125rem;
  background-color: rgb(225,230,243);
  color: rgb(62,67,87);
}

<div>
  <div
    class="sw-flex sw-gap-1"
  >
    <span
      class=" emotion-0 emotion-1"
    >
      <svg
        aria-hidden="true"
        class="octicon octicon-triangle-up"
        fill="currentColor"
        focusable="false"
        height="16"
        style="display: inline-block; user-select: none; vertical-align: middle; overflow: visible;"
        viewBox="0 0 16 16"
        width="16"
      >
        <path
          d="m4.427 9.573 3.396-3.396a.25.25 0 0 1 .354 0l3.396 3.396a.25.25 0 0 1-.177.427H4.604a.25.25 0 0 1-.177-.427Z"
        />
      </svg>
    </span>
  </div>
</div>
`;

exports[`should render Click 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  padding-left: 0.125rem;
  padding-right: 0.125rem;
  border-radius: 0.125rem;
  background-color: rgb(225,230,243);
  color: rgb(62,67,87);
}

<div>
  <div
    class="sw-flex sw-gap-1"
  >
    <span
      class=" emotion-0 emotion-1"
    >
      click
    </span>
  </div>
</div>
`;

exports[`should render Command 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  padding-left: 0.125rem;
  padding-right: 0.125rem;
  border-radius: 0.125rem;
  background-color: rgb(225,230,243);
  color: rgb(62,67,87);
}

<div>
  <div
    class="sw-flex sw-gap-1"
  >
    <span
      class=" emotion-0 emotion-1"
    >
      ⌘
    </span>
  </div>
</div>
`;

exports[`should render Control 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  padding-left: 0.125rem;
  padding-right: 0.125rem;
  border-radius: 0.125rem;
  background-color: rgb(225,230,243);
  color: rgb(62,67,87);
}

<div>
  <div
    class="sw-flex sw-gap-1"
  >
    <span
      class=" emotion-0 emotion-1"
    >
      Ctrl
    </span>
  </div>
</div>
`;

exports[`should render Option 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  padding-left: 0.125rem;
  padding-right: 0.125rem;
  border-radius: 0.125rem;
  background-color: rgb(225,230,243);
  color: rgb(62,67,87);
}

<div>
  <div
    class="sw-flex sw-gap-1"
  >
    <span
      class=" emotion-0 emotion-1"
    >
      ⌥
    </span>
  </div>
</div>
`;

exports[`should render a default text if no keys match 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  padding-left: 0.125rem;
  padding-right: 0.125rem;
  border-radius: 0.125rem;
  background-color: rgb(225,230,243);
  color: rgb(62,67,87);
}

<div>
  <div
    class="sw-flex sw-gap-1"
  >
    <span
      class=" emotion-0 emotion-1"
    >
      Ctrl
    </span>
    <span>
      +
    </span>
    <span
      class=" emotion-0 emotion-1"
    >
      click
    </span>
  </div>
</div>
`;

exports[`should render multiple keys 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  padding-left: 0.125rem;
  padding-right: 0.125rem;
  border-radius: 0.125rem;
  background-color: rgb(225,230,243);
  color: rgb(62,67,87);
}

<div>
  <div
    class="sw-flex sw-gap-1"
  >
    <span
      class="sw-px-1 emotion-0 emotion-1"
    >
      Use
    </span>
    <span
      class=" emotion-0 emotion-1"
    >
      Ctrl
    </span>
    <span>
      +
    </span>
    <span
      class=" emotion-0 emotion-1"
    >
      <svg
        aria-hidden="true"
        class="octicon octicon-triangle-up"
        fill="currentColor"
        focusable="false"
        height="16"
        style="display: inline-block; user-select: none; vertical-align: middle; overflow: visible;"
        viewBox="0 0 16 16"
        width="16"
      >
        <path
          d="m4.427 9.573 3.396-3.396a.25.25 0 0 1 .354 0l3.396 3.396a.25.25 0 0 1-.177.427H4.604a.25.25 0 0 1-.177-.427Z"
        />
      </svg>
    </span>
    <span
      class=" emotion-0 emotion-1"
    >
      <svg
        aria-hidden="true"
        class="octicon octicon-triangle-down"
        fill="currentColor"
        focusable="false"
        height="16"
        style="display: inline-block; user-select: none; vertical-align: middle; overflow: visible;"
        viewBox="0 0 16 16"
        width="16"
      >
        <path
          d="m4.427 7.427 3.396 3.396a.25.25 0 0 0 .354 0l3.396-3.396A.25.25 0 0 0 11.396 7H4.604a.25.25 0 0 0-.177.427Z"
        />
      </svg>
    </span>
  </div>
</div>
`;

exports[`should render multiple keys with non-key symbols 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  padding-left: 0.125rem;
  padding-right: 0.125rem;
  border-radius: 0.125rem;
  background-color: rgb(225,230,243);
  color: rgb(62,67,87);
}

<div>
  <div
    class="sw-flex sw-gap-1"
  >
    <span
      class=" emotion-0 emotion-1"
    >
      Ctrl
    </span>
    <span>
      +
    </span>
    <span
      class=" emotion-0 emotion-1"
    >
      <svg
        aria-hidden="true"
        class="octicon octicon-triangle-down"
        fill="currentColor"
        focusable="false"
        height="16"
        style="display: inline-block; user-select: none; vertical-align: middle; overflow: visible;"
        viewBox="0 0 16 16"
        width="16"
      >
        <path
          d="m4.427 7.427 3.396 3.396a.25.25 0 0 0 .354 0l3.396-3.396A.25.25 0 0 0 11.396 7H4.604a.25.25 0 0 0-.177.427Z"
        />
      </svg>
    </span>
    <span
      class=" emotion-0 emotion-1"
    >
      <svg
        aria-hidden="true"
        class="octicon octicon-triangle-up"
        fill="currentColor"
        focusable="false"
        height="16"
        style="display: inline-block; user-select: none; vertical-align: middle; overflow: visible;"
        viewBox="0 0 16 16"
        width="16"
      >
        <path
          d="m4.427 9.573 3.396-3.396a.25.25 0 0 1 .354 0l3.396 3.396a.25.25 0 0 1-.177.427H4.604a.25.25 0 0 1-.177-.427Z"
        />
      </svg>
    </span>
  </div>
</div>
`;
